src/azul/service/hca_response_v5.py

```diff
@@ -317,107 +313,23 @@ def make_projects(self, entry):
-
-        files = [
-            {
-                # Each line in the stratification string represents a stratum,
-                # each stratum is a list of points, each point has a dimension
-                # and a list of values. Transform that string into a list of
-                # dictionaries. Each entry in those dictionaries maps the
-                # dimension to a value in that dimension. If dimension in a
-                # stratum has multiple values, the stratum is expanded into
-                # multiple strata, one per value. The strata are identical
-                # except in the dimension that had the multiple values.
-                'strata': list(chain.from_iterable(
-                    map(dict, product(*(
-                        [(dimension, value) for value in values.split(',')]
-                        for dimension, values in (point.split('=') for point in stratum.split(';'))
-                    )))
-                    for stratum in file['strata'].split('\n')
-                )),
-                'url': furl(config.service_endpoint(),
-                            path=('fetch', 'repository', 'files', file['uuid']),
-                            args=dict(version=file['version'],
-                                      catalog=catalog))
-            }
-            for file in files
-        ]
-
-        # To produce a tree with the most shared base branches possible we sort
-        # the dimensions by number of distinct values on that dimension.
-        distinct_values = defaultdict(set)
-        for file in files:
-            for stratum in file['strata']:
-                for dimension, value in stratum.items():
-                    distinct_values[dimension].add(value)
-        sorted_dimensions = sorted(distinct_values,
-                                   key=lambda k: len(distinct_values[k]))
-
-        # Build the tree, as a nested dictionary. The keys in the dictionary
-        # alternate between dimensions and values. The leaves of the tree are
-        # lists of matrix file URLs. If a matrix covers multiple strata, its
-        # URL will occur multiple times in the tree.
-        tree = NestedDict(2 * len(sorted_dimensions), list)
-        for file in files:
-            for stratum in file['strata']:
-                node = tree
-                for dimension in sorted_dimensions:
-                    value = stratum.get(dimension)
-                    if value is not None:
-                        node = node[dimension][value]
-                node['url'].append(file['url'].url)
-
-        return tree
+        return make_contributor_matrices_tree(self.catalog, files, self.make_file_url)
+
+    def make_file_url(self,
+                      catalog: CatalogName,
+                      file_id: str,
+                      file_version: str) -> str:
+        return furl(config.service_endpoint(),
+                    path=('fetch', 'repository', 'files', file_id),
```

Comment:
This is still brittle. Only the app module and maybe a controller should be aware of request routing. When we write code like this, we need to be aware of the dependencies it introduces, if they are brittle and what the failure mode would be if they broke. If someone came along and modified the route in app.py, they would in all likelihood miss to update this. We wouldn't learn about the breakage until a user actually tried to access this URL. An integration test might catch this but we have none for this right now and often they aren't run until a PR is tested in the sandbox.

And even if we had an IT, this violates the principle of separation of concerns and make future maintenance of this code harder.